### ISSUE TITLE
[DI] Compile Error on cache warming since Symfony 3.4 if cache warmed on different machine already (host vs. VM)

### ISSUE THREAD

**Issue by jlekowski** [symfony/symfony#25857]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | yes
| RFC?             | no
| Symfony version  | >=3.4

To reproduce you need a host machine and VM/Docker:
1. On host run:
```
$ bin/console cache:clear --no-warmup -vv --env=dev && bin/console cache:warmup -vv --env=dev
```
2. On VM/Docker run:
```
$ bin/console
```
3. On host run:
```
$ bin/console

PHP Warning:  Uncaught ErrorException: Warning: require({removed_path}/var/cache/dev/ContainerDbj4Wfg/getCacheWarmerService.php): failed to open stream: No such file or directory in {removed_path}/var/cache/dev/ContainerDbj4Wfg/srcDevDebugProjectContainer.php:353
Stack trace:
#0 {removed_path}/var/cache/dev/ContainerDbj4Wfg/srcDevDebugProjectContainer.php(353): require()
#1 {removed_path}/vendor/symfony/dependency-injection/Container.php(232): ContainerDbj4Wfg\srcDevDebugProjectContainer->load('/home/jerzy/dev...')
#2 {removed_path}/vendor/symfony/http-kernel/Kernel.php(543): Symfony\Component\DependencyInjection\Container->get('cache_warmer')
#3 {removed_path}/vendor/symfony/http-kernel/Kernel.php(125): Symfony\Component\HttpKernel\Kernel->initializeContainer()
#4 {removed_path}/vendor/symfony/framework-bundle/Console/Application.php(63): Symfony\Component\HttpKer in {removed_path}/var/cache/dev/ContainerDbj4Wfg/srcDevDebugProjectContainer.php on line 353
PHP Fatal error:  ContainerDbj4Wfg\srcDevDebugProjectContainer::load(): Failed opening required '{removed_path}/var/cache/dev/ContainerDbj4Wfg/getCacheWarmerService.php' (include_path='.:/usr/share/php') in {removed_path}/var/cache/dev/ContainerDbj4Wfg/srcDevDebugProjectContainer.php on line 353

In srcDevDebugProjectContainer.php line 353:
                                                                                                                                                                                                                  
  Compile Error: ContainerDbj4Wfg\srcDevDebugProjectContainer::load(): Failed opening required '{removed_path}/var/cache/dev/ContainerDbj4Wfg/getCacheWarmerService.php' (include_path='.  
  :/usr/share/php')
```
4. On host run again:
```
$ bin/console

PHP Warning:  include({removed_path}/var/cache/dev/ContainerDbj4Wfg/srcDevDebugProjectContainer.php): failed to open stream: No such file or directory in {removed_path}/var/cache/dev/srcDevDebugProjectContainer.php on line 7
PHP Warning:  include(): Failed opening '{removed_path}/var/cache/dev/ContainerDbj4Wfg/srcDevDebugProjectContainer.php' for inclusion (include_path='.:/usr/share/php') in {removed_path}/var/cache/dev/srcDevDebugProjectContainer.php on line 7
Symfony 4.0.3 (kernel: src, env: dev, debug: true)
```
But this time, besides the warning, I get correct list of all command.

5. On host run again:
```
$ bin/console
```
Working as expected.

The issue is reproducible also when warming up cache on VM/Docker first, running command on host, and accessing VM/Docker through web. First web request throws an error, second is OK.

In `3.3` there were no `Container{$hash}` (generated here https://github.com/symfony/dependency-injection/blob/master/Dumper/PhpDumper.php#L212) folders in cache, since `3.4` there is one, and hash generated on host and VM/Docker is different. Here is a link to some debugging I did: https://gist.github.com/jlekowski/931fd71e06655bfa45a7ac0200e8051b

**Comment by nicolas-grekas** [comment]
Here are some hints to help you debug the situation:
- the hash is computed by the content of the ContainerHashhh/ folder. you should compare both folders (eg using `meld`) and spot the differences. Are they legitimate? A container should be stable if the config doesn't change.
- the `var/cache/dev/srcDevDebugProjectContainer.php` should always reference the current hash. Why isn't that the case on the first failing step?

**Comment by jlekowski** [comment]
@nicolas-grekas, here is the diff between folders https://gist.github.com/jlekowski/931fd71e06655bfa45a7ac0200e8051b#file-diff-log After creating them on each environment, I copied them to make sure another warmup will not change anything in there.

In general, files generated on host machine contain lines like (e.g. `getAnnotations_CacheWarmerService.php`):
```
include_once $this->targetDirs[3].'/vendor/symfony/http-kernel/CacheWarmer/CacheWarmerInterface.php';
include_once $this->targetDirs[3].'/vendor/symfony/framework-bundle/CacheWarmer/AbstractPhpFileCacheWarmer.php';
include_once $this->targetDirs[3].'/vendor/symfony/framework-bundle/CacheWarmer/AnnotationsCacheWarmer.php';
```

After run on each env `var/cache/dev/srcDevDebugProjectContainer.php` changes and points to the right hash.

Before step 3 `var/cache/dev/ContainerDbj4Wfg/getCacheWarmerService.php` file exists (host hash), as it was created on first cache warmup on host. When I run `bin/console` it is removed during the process and error occurs. After I get the error, the file does not exist any more. It is recreated in step 4.

**Comment by nicolas-grekas** [comment]
Fixed in #25866

**Comment by nicolas-grekas** [comment]
Hum, #25866 fixes the missing `include_once` in Docker, but the hash will still be different between host/container because of the way the cache's namespace is generated. If you want the same has, you need to set the `cache.prefix.seed` parameter to the unique name of your app (same value as what you'd put in the "name" field of your composer.json.)

Before doing so, would you mind sharing a reproducer, eg a Dockerfile + precise instructions/commands to hit the error?

**Comment by jlekowski** [comment]
I applied #25866, but no changes.

To reproduce, of course you need Symfony:
```
composer create-project "symfony/skeleton" demo
cd demo
```

I use `docker-compose`, so create `docker-compose.yml`:
```
version: '2'
services:
  web:
    container_name: demo
    build: .
    volumes:
      - ./:/app
    tty: true
```
Create `Dockerfile`:
```
FROM amazonlinux:latest
RUN yum install -y php71 php71-intl php71-mbstring php71-pecl-apcu php71-opcache
WORKDIR /app
```

Now run:
```
docker-compose up -d
bin/console
docker exec -t -i demo /app/bin/console
bin/console
```

**Comment by nicolas-grekas** [comment]
Thanks for the reproducer. Issue has already been fixed in #25733. Fix will be part of next release.

**Comment by jlekowski** [comment]
Thanks for looking at it Nicolas, #25733 fixes the issue.
Good speech in Cluj :)
